File: src/scraper.py
"""
Fetches current stock prices from NEPSE data sources.
Primary:  financialnotices.com (confirmed working)
Fallback: merolagani.com
"""

import re
import time
import requests
from bs4 import BeautifulSoup

HEADERS = {
    "User-Agent": (
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
        "AppleWebKit/537.36 (KHTML, like Gecko) "
        "Chrome/122.0.0.0 Safari/537.36"
    ),
    "Accept-Language": "en-US,en;q=0.9",
}
TIMEOUT = 20


def _parse_financialnotices(symbol: str) -> float | None:
    """Scrape current price from financialnotices.com."""
    url = f"https://www.financialnotices.com/stock-nepse.php?symbol={symbol}"
    try:
        r = requests.get(url, headers=HEADERS, timeout=TIMEOUT)
        r.raise_for_status()

        # Pattern: "NABIL was closed at RS. 543 on 2026-03-26"
        match = re.search(
            r'closed at RS\.\s*([\d,]+(?:\.\d+)?)',
            r.text, re.IGNORECASE
        )
        if match:
            return float(match.group(1).replace(",", ""))

        # Fallback: look for price in structured data tags
        soup = BeautifulSoup(r.text, "html.parser")
        for tag in soup.find_all(string=re.compile(r'RS\.\s*[\d,]+')):
            m = re.search(r'RS\.\s*([\d,]+(?:\.\d+)?)', tag, re.IGNORECASE)
            if m:
                return float(m.group(1).replace(",", ""))
    except Exception as e:
        print(f"[scraper] financialnotices failed for {symbol}: {e}")
    return None


def _parse_merolagani(symbol: str) -> float | None:
    """Fallback: scrape last traded price from merolagani.com."""
    url = f"https://merolagani.com/CompanyDetail.aspx?symbol={symbol}"
    try:
        r = requests.get(url, headers=HEADERS, timeout=TIMEOUT)
        r.raise_for_status()
        soup = BeautifulSoup(r.text, "html.parser")

        # Merolagani shows LTP in a span with class 'ltp' or similar
        for candidate in ["ltp", "price", "market-price"]:
            el = soup.find(class_=re.compile(candidate, re.IGNORECASE))
            if el:
                m = re.search(r'[\d,]+(?:\.\d+)?', el.get_text())
                if m:
                    return float(m.group().replace(",", ""))
    except Exception as e:
        print(f"[scraper] merolagani failed for {symbol}: {e}")
    return None


def fetch_price(symbol: str) -> float | None:
    """
    Returns current LTP (Last Traded Price) for the given NEPSE symbol.
    Tries primary source first, falls back to secondary.
    """
    price = _parse_financialnotices(symbol)
    if price:
        return price

    print(f"[scraper] Primary failed for {symbol}, trying fallback...")
    time.sleep(2)
    return _parse_merolagani(symbol)


def fetch_all_prices(symbols: list[str]) -> dict[str, float | None]:
    """
    Fetches prices for all symbols with a small delay between requests
    to avoid hammering the server.
    """
    results = {}
    for i, symbol in enumerate(symbols):
        results[symbol] = fetch_price(symbol)
        print(f"[scraper] {symbol}: {results[symbol]}")
        if i < len(symbols) - 1:
            time.sleep(2)  # polite delay between requests
    return results

File: worker/src/scraper.js
/**
 * NEPSE price scraper — fetches latest traded price for a stock symbol.
 * Primary source: financialnotices.com
 * Fallback source: merolagani.com
 */

const UA = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36';

async function fetchFromFinancialNotices(symbol) {
  const url = `https://www.financialnotices.com/stock-nepse.php?symbol=${symbol}`;
  const res = await fetch(url, { headers: { 'User-Agent': UA }, cf: { cacheTtl: 0 } });
  if (!res.ok) return null;
  const html = await res.text();

  // "NABIL was closed at RS. 543.00 on 2026-03-31"
  let m = html.match(/was\s+closed\s+at\s+RS\.\s*([\d,]+\.?\d*)/i);
  if (m) return parseFloat(m[1].replace(/,/g, ''));

  // fallback pattern: RS. <number>
  m = html.match(/RS\.\s*([\d,]+\.?\d*)/i);
  if (m) return parseFloat(m[1].replace(/,/g, ''));

  return null;
}

async function fetchFromMerolagani(symbol) {
  const url = `https://merolagani.com/CompanyDetail.aspx?symbol=${symbol}`;
  const res = await fetch(url, { headers: { 'User-Agent': UA }, cf: { cacheTtl: 0 } });
  if (!res.ok) return null;
  const html = await res.text();

  // Look for LTP value in common class patterns
  for (const pattern of [
    /class="ltp[^"]*"[^>]*>\s*([\d,]+\.?\d*)/i,
    /class="[^"]*price[^"]*"[^>]*>\s*([\d,]+\.?\d*)/i,
    /"market-price"[^>]*>\s*([\d,]+\.?\d*)/i,
    /Last\s+Traded\s+Price[^<]*<[^>]+>\s*([\d,]+\.?\d*)/i,
  ]) {
    const m = html.match(pattern);
    if (m) return parseFloat(m[1].replace(/,/g, ''));
  }

  return null;
}

/**
 * Fetch the latest price for a single symbol.
 * Returns the price as a number, or null if both sources fail.
 */
export async function fetchPrice(symbol) {
  try {
    const price = await fetchFromFinancialNotices(symbol);
    if (price !== null && price > 0) return price;
  } catch (_) {}

  try {
    const price = await fetchFromMerolagani(symbol);
    if (price !== null && price > 0) return price;
  } catch (_) {}

  return null;
}

/**
 * Fetch prices for multiple symbols concurrently (with a small delay between
 * batches to avoid hammering the source sites).
 * Returns { SYMBOL: price|null, ... }
 */
export async function fetchAllPrices(symbols) {
  const results = {};
  // Fetch in batches of 3 with a 1-second gap between batches
  for (let i = 0; i < symbols.length; i += 3) {
    const batch = symbols.slice(i, i + 3);
    const prices = await Promise.all(batch.map(s => fetchPrice(s)));
    batch.forEach((s, idx) => { results[s] = prices[idx]; });
    if (i + 3 < symbols.length) {
      await new Promise(r => setTimeout(r, 1000));
    }
  }
  return results;
}
